scripts/_common.sh

```diff
@@ -130,3 +126,60 @@ function use_gateway_binaries_for_version() {
 }
 export -f use_gateway_binaries_for_version
 
+function generate_matrix() {
+  filter_fn="$1"
+  shift
+
+  versions=("$@")
+  version_matrix=()
+  for fed_version in "${versions[@]}"; do
+    for client_version in "${versions[@]}"; do
+      for gateway_version in "${versions[@]}"; do
+        if "$filter_fn" "$fed_version" "$client_version" "$gateway_version"; then
+          continue
+        fi
+
+        version_matrix+=("$fed_version $client_version $gateway_version")
+      done
+    done
+  done
+
+  for version in "${version_matrix[@]}"; do
+    echo "$version"
+  done
+}
+
+function full_matrix_filter() {
```

Comment:
Not clear what and why it does. Comment would be helpful. Also should it be called `is_...` ?

Reply:
Let me know if the refactors for an inclusive filter plus the comments/docs clears this up (or if it's too verbose). Happy to keep iterating.

Reply:
I got more comments than I could dream for. :)